GUI Support for Enabling/Disabling Native Messaging Support Per Browser

eajaykumar opened the pull request:
## Summary

This PR adds a new **Browser Integration** tab in the Persepolis settings, giving users control over which browsers are integrated via native messaging.

### Features:
1. Users can now explicitly select browsers (Firefox, LibreWolf, Chrome, etc.) to integrate with Persepolis.
2. When a browser is unselected, its corresponding native messaging manifest is automatically removed.
3. Manifests are no longer created unnecessarily. Integration is now **opt-in**.
4. The list of browsers is shown even if not installed (but auto-check only if installed).
5. GUI checkboxes reflect real integration status.
6. Custom manifest path is supported.
- Changes made to: `initialization.py`, `setting_ui.py`, `setting.py`, `browser_integration.py`


 Tested on Linux (LibreWolf, Firefox). Requires clean config reset for testing.

eajaykumar committed:
Add configurable browser integration tab with native messaging manifest management

eajaykumar committed:
clean debug code

eajaykumar commented:
@alireza-amirsamimi I've done my best to keep the code clean, modular. However, feel free to edit, simplify, or remove any parts of the code if you find something unnecessary or if it can be done better.
Thanks  with this i can say any custom Browser and officials ones will work smoothly with persepolis

alireza-amirsamimi commented:
@eajaykumar 
Thank you. I'll review and accept your pull request soon :+1:

ghost commented:
When testing this branch on Windows, I encountered the following errors:

```log
Failed to remove manifest for librewolf: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for firefox: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for vivaldi: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for opera: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for chrome: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for brave: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for chromium: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
```

It seems the issue is that `shutil.which()` fails to find any browsers on Windows, as they are not typically added to the system `PATH`. This results in `enabled_browsers` being an empty list.

Consequently, `remove_manifests_for_browsers` is called for all `all_known_browsers,` but since `get_native_message_folder` doesn't have a Windows implementation, it returns None, leading to the `TypeError: path should be string... not NoneType`.

This suggests that relying on `shutil.which()` for browser detection is not enough for cross-platform support. Maybe we could using a more specialized library like [browsers](https://github.com/roniemartinez/browsers) to handle this?

eajaykumar commented:
@guest75582332 ,its good way to use cross-platform library browsers ,thanks for testing and recommendation as i don't have windows or mac so i didn't test with those platforms ,i think its a one way shot to support all platforms now with this library i will try and implement it

eajaykumar commented:
i checked the library its only detecting official browsers not custom browsers. can we fork [browsers](https://github.com/roniemartinez/browsers) library to support well know custom browsers. i need final call from @alireza-amirsamimi to make it happen

alireza-amirsamimi commented:
Hello,

I have read your code. Thank you very much for your efforts. 

My concern is that the code should work across all operating systems and, of course, not be overly complex so that we can maintain it easily in the future. 

I believe it is a good strategy for Persepolis to create configuration files only for Chrome, Chromium, and Firefox. Other common browsers(Opera, Brave, Vivaldi, Libre Wolf) can be selected by the user in the settings window if needed. 

We must acknowledge that there are many browsers, and their number is increasing every day, but in reality, users do not use them as their primary browser and there aren't many users for them. I think there is no reason to complicate the code for these browsers. What I mean is that there is no need for a custom path section in the settings window. 

If users of these browsers are Persepolis users, they will likely either submit a pull request or open an issue, and we can add the required browser to the code in the future.

eajaykumar commented:
OK its a good strategy so we remove custom manifest path and Only Firefox, Chrome, and Chromium are enabled by default. Users have Options can toggle additional browsers (LibreWolf, Brave, etc.) via the GUI. and we can add support to other browsers per request of users. so no forking of browsers library is required but should we add browsers library for cross platform support or do we hard-code those three default browsers for detecting browser installation. 
and also one suggestion is that we need to inform users about browser integration to those who use other custom browsers that wont automatically work by default maybe we should add a line in web extension to inform that there is a browser integration tab that exists to enable them

ghost commented:
One of the original intentions behind proposing a custom path option was to support non-standard variants (e.g., Chrome Beta, Chrome Dev, Chrome Canary). In my opinion, you might not want to explicitly support these variants. That said, since all code maintenance ultimately falls on the maintainer, it's still workable without a custom path option—copying Chrome's manifest file to the appropriate location still works. As long as there's a way to clean up useless manifest files, I think that should be sufficient.

eajaykumar commented:
@guest75582332  You're right copying Chrome's manifest manually for other variants  works just fine. keeping the implementation simple, and focusing on core browsers like Firefox, Chrome, and Chromium, makes long-term maintenance easier. unused manifest files are properly cleaned up while user unchecks the browser, so users won't be left with unnecessary configurations.

eajaykumar commented:
i just need a final go and advice from @alireza-amirsamimi  to add browsers library to use as dependency and also its very light weight  as i saw the code. i will remove custom manifest path and only Firefox, Chrome, and Chromium are enabled defaults.

alireza-amirsamimi commented:
> i just need a final go and advice from @alireza-amirsamimi to add browsers library to use as dependency and also its very light weight as i saw the code. i will remove custom manifest path and only Firefox, Chrome, and Chromium are enabled defaults.

What do you mean by the browser library?

eajaykumar commented:
for detecting installed browsers in Linux, macs, windows. [browsers](https://github.com/roniemartinez/browsers)
with this we can only toggle the check for installed default browsers or we have one option we can hard code enable them

alireza-amirsamimi commented:
@eajaykumar 
We can only use libraries that are well-known and are available in the official repositories of all Linux distributions and FreeBSD, such as requests, urllib3, psutil, ... . Otherwise, those who maintain the Perspolis package in the distributions' repositories will not be able to update it.

ghost commented:
I checked the status of the `pybrowsers` library on Repology, it is not a widely packaged library, which would certainly complicate things for downstream maintainers.

https://repology.org/project/pybrowsers/versions

Given this, perhaps the best approach for now is to implement the browser detection logic internally. We could hard-code the necessary checks for the primary targets (Chrome, Chromium, and Firefox) across Windows, macOS, and Linux.

eajaykumar committed:
cross platform support added for detecting browsers installed in linux ,mac ,windows

eajaykumar commented:
@guest75582332 can you test it and update result please ,i added support for windows to detect installed browsers

alireza-amirsamimi commented:
So we cannot use the browser library because it is not available in the repositories of Arch Linux, Fedora, openSUSE, FreeBSD, and so on. My ten years of experience in software development shows that package maintainers are reluctant to add a new package that is not in the official distribution repositories, as the process of adding and accepting a package in well-known distributions involves many steps. This means that if we use this library in the code, it is likely that Persepolis will not be updated in these well-known distributions.

eajaykumar commented:
yea i too recognize it ,now i dropped it and hard coded it to streamline every step smoothly without using browsers library, you can check the code

eajaykumar committed:
macos browser detection fix

ghost commented:
This does fix the browser detection issue. However, `remove_manifests_for_browsers` still isn’t correctly implemented for non-Linux platforms, which leads to the following errors:

```log
Failed to remove manifest for librewolf: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for brave: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for chromium: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for opera: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for vivaldi: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
Failed to remove manifest for firefox: _path_exists: path should be string, bytes, os.PathLike or integer, not NoneType
```

eajaykumar commented:
@guest75582332 i will install virtual windows 10 and will try myself thanks. so its showing default browser check boxes and can you check in browsers folders if manifests generated properly

ghost commented:
```python
def remove_manifests_for_browsers(browsers, custom_path=None):
    for browser in browsers:
        try:
            if browser.strip() == "":
                continue
            path = get_manifest_path_for_browser(browser, custom_path)
            if os.path.exists(path):
                os.remove(path)
                print(f"Removed manifest for {browser}: {path}")
        except Exception as e:
            print(f"Failed to remove manifest for {browser}: {e}")
```

I think the reason for the error is that the current implementation of `remove_manifests_for_browsers` relies on the file paths returned by `get_manifest_path_for_browser`, which is currently only implemented for Linux: 

```python 
if os_type == OS.LINUX: 
 if browser == BROWSER. CHROMIUM: 
 return os.path.join(home_address, '.config/chromium/NativeMessagingHosts') 
 elif browser == BROWSER.CHROME: 
 return os.path.join(home_address, '.config/google-chrome/NativeMessagingHosts') 
 elif browser == BROWSER.FIREFOX: 
 return os.path. join(home_address, '.mozilla/native-messaging-hosts') 
 elif browser == BROWSER.VIVALDI: 
 return os.path.join(home_address, '.config/vivaldi/NativeMessagingHosts') 
 elif browser == BROWSER.OPERA: 
 return os.path.join(home_address, '. config/opera/NativeMessagingHosts') 
 elif browser == BROWSER.BRAVE: 
 return os.path.join(home_address, '.config/BraveSoftware/Brave-Browser/NativeMessagingHosts') 
 elif browser == "librewolf": 
 return os.path.join(home_address, '.librewolf/native-messaging-hosts') 
```

eajaykumar commented:
@guest75582332 yea i got it that's why am installing windows and checking all browsers paths to implement and make sure everything work properly

eajaykumar committed:
add manifest path for windows and mac

eajaykumar commented:
i got it working in windows now ,am just refactoring code and will update it

eajaykumar commented:
@alireza-amirsamimi @guest75582332 i encountered conflict on Windows, all Chromium-based browsers (Chrome, Brave, Chromium,etc.) shared the same native messaging host name and manifest. This caused a conflict where disabled browsers could still capture downloads, because Windows uses the same registry key for all as the extension connects to the same host name and extension id. 
now i added unique native messaging host names for each Chromium-based browser like  com.persepolis.chrome, com.persepolis.brave etc. updated the manifest generation and registry entries so that each browser uses its own native host. this prevents disabled Chromium browsers from capturing downloads when not selected in settings. 
to make this work i need to edit code in persepolis web extension to detect its own host name  
```
function getNativeHostName() {
    const browser = detectBrowser();
    if (DEBUG) {
        console.log("Detected browser:", browser);
    }
    switch (browser) {
        case "chrome":
            return "com.persepolis.chrome";
        case "brave":
            return "com.persepolis.brave";
        case "edge":
            return "com.persepolis.edge";
        case "vivaldi":
            return "com.persepolis.vivaldi";
        default:
            return "com.persepolis.pdmchromewrapper";
    }
}
```
with this changes only the enabled browser’s manifest works and every browser has unique host name  and we can add a fallback to for older version with this `com.persepolis.pdmchromewrapper`, with this changes its efficiently captures links

alireza-amirsamimi commented:
@eajaykumar 
Thank you for the effort you put in. Let me see if I can find an easier solution.

eajaykumar commented:
@alireza-amirsamimi fallback logic not supported for vivaldi ,chroimum in browser extension so it will break persepolis extenstion for previous versions of download manager , i have tested for firefox ,librewolf,chrome and brave if we need to implement enabling and disabling per broswer then everyone who uses persepolis must update to latest version 

actually they dont have any apis to get detected as vivaldi or chromium they give output as chrome so it will not work for fallback

eajaykumar committed:
enable disable browser integration with host names

eajaykumar commented:
[updated web extension](https://github.com/persepolisdm/Persepolis-WebExtension/commit/3d8bf771e3d98eeefc4e27509936e4df8912764e) you can check and make a decision about fallback to com.persepolis.pdmchromewrapper or for easier solution will just hard code it to web extension but it will break previous persepolis versions

alireza-amirsamimi commented:
@eajaykumar 
If it will break previous versions so It's not good choice. In my opinion, we should deal with the matter a little more simply and not try to solve one problem and start the next problems. We can solve the important part and leave the other part. I was a little busy. I will complete the code you wrote soon and merge the pull request.

eajaykumar commented:
@alireza-amirsamimi after a deep thinking and some sleep ,i got quite an extraordinary smart way to implement unique  manifest for chromium based browsers even if we cant detect the browser and this logic should apply for all firefox based browsers too,now it solves web extension problem for fallback to previous version manifest file and now we can proceed to implement enabling and disabling native messaging per browser

eajaykumar commented:
for windows its not working as registry key try only works with chrome ,chromium browser not others 
```
 BROWSER.CHROME: r"SOFTWARE\Google\Chrome\NativeMessagingHosts\com.persepolis.chrome",
    BROWSER.CHROMIUM: r"SOFTWARE\Chromium\NativeMessagingHosts\com.persepolis.chromium",
    BROWSER.BRAVE: r"SOFTWARE\BraveSoftware\Brave-Browser\NativeMessagingHosts\com.persepolis.brave",
    BROWSER.VIVALDI: r"SOFTWARE\Vivaldi\NativeMessagingHosts\com.persepolis.vivaldi",
    BROWSER.OPERA: r"SOFTWARE\Opera Software\Opera Stable\NativeMessagingHosts\com.persepolis.opera",
    BROWSER.FIREFOX: r"SOFTWARE\Mozilla\NativeMessagingHosts\com.persepolis.firefox",
    "librewolf": r"SOFTWARE\Mozilla\NativeMessagingHosts\com.persepolis.librewolf"
```
this will will work for chrome and chromium as brave,opera,vivaldi will not commuicate with their own native messaging paths 

```
BROWSER.CHROME: r"SOFTWARE\Google\Chrome\NativeMessagingHosts\com.persepolis.chrome",
    BROWSER.CHROMIUM: r"SOFTWARE\Chromium\NativeMessagingHosts\com.persepolis.chromium",
    BROWSER.BRAVE: r"SOFTWARE\Google\Chrome\NativeMessagingHosts\com.persepolis.brave",
    BROWSER.VIVALDI: r"SOFTWARE\Google\Chrome\NativeMessagingHosts\com.persepolis.vivaldi",
    BROWSER.OPERA: r"SOFTWARE\Google\Chrome\NativeMessagingHosts\com.persepolis.opera",
    BROWSER.FIREFOX: r"SOFTWARE\Mozilla\NativeMessagingHosts\com.persepolis.firefox",
    "librewolf": r"SOFTWARE\Mozilla\NativeMessagingHosts\com.persepolis.librewolf"
```

 this will work as registry key points to chrome registry key ,i dont know why browsers cant point to its own native messaging registry key ,with this implementation all browsers work except vivaldi as it dont have specific apis to detect browsers

we have two options as of now
1. just merge chrome/Vivaldi as one check box just for windows but i think its not a good way of implementation 
2. implement every browser with unique extension id (what i mean is separate extension for every browser like  Persepolis Download Manager Integration (chrome) (brave) (Vivaldi) )  more maintenance if we implement this, but its safe and efficient way
i tried all my best in implementation but this crazy limitation of windows that every chromium based browser should only register native messaging key within chrome browser just blown away all hopes

ghost commented:
Option 1 is functionally sufficient, since the idea of being able to enable or disable Native Messaging is mainly intended for users who don't want to see unnecessary files. I believe it's rare for someone to install a browser extension but want it to be non-functional, so this might be acceptable—even if it's conceptually a bit odd.

eajaykumar commented:
actually now am getting sense that just making sure every browser works with download manager and capturing links is enough,and supporting all popular custom browsers by user requests and  we can fix unnecessary files just by detecting which browsers installed and only installing manifests for those . this enable and disable thing only possible if we have extension for every browser reliably 
@guest75582332 which custom browsers you use so i will make sure it works out of box without browser integration ?

ghost commented:
I am using Chrome Dev (a non-trivial variant of a common browser). Based on our previous discussion, no special handling is required — we just need to ensure that we don’t generate unnecessary files for the user. On Windows, Chromium-based browsers effectively share registry entries to some extent, so there's nothing redundant, and therefore this issue does not arise.

eajaykumar commented:
@guest75582332 essentially with one registry key and one manifest file in chrome will work for all chromium based browsers even custom ones, its just how windows works "**One key to rule them all, One manifest to find them, One Browser to bring them all and in the darkness that binds them**" 🧐😂😂😂

alireza-amirsamimi commented:
Hello, @eajaykumar @guest75582332 
Thank you for your valuable effort.
I made some changes to your code and opened a new pull request for it [here](https://github.com/persepolisdm/persepolis/pull/1095).
Therefore, I will close this pull request.
